Feature/vj-notebook

VarenyaJ opened the pull request:
Refactor project; start notebook from scratch

SmartMonkey-git committed:
Push notebook

SmartMonkey-git committed:
Notebook update

VarenyaJ committed:
Merge remote-tracking branch 'origin/rr/model-notebook' into feature/experiment-notebook
Starting new work on the Trello ticket, continuing work started in "rr/model-notebook"

VarenyaJ committed:
Update README.md from main

VarenyaJ committed:
Installing jupyterlab via pip

VarenyaJ committed:
start setting up jupyter notebook

VarenyaJ committed:
Update README.md from main

VarenyaJ committed:
Add warning to the scripts README

VarenyaJ committed:
Ignore Mac-specific files generated by Apple

VarenyaJ committed:
Adding .idea/ files

VarenyaJ committed:
Trying to work on setting up the python notebook but cannot generate 'P5/scripts/data/tmp/phenopacket_dataset.csv' from 'P5/scripts/readme.md'

VarenyaJ committed:
Fix Pycharm to show Project Root

VarenyaJ committed:
Clean up imports

VarenyaJ committed:
Writing things a little out of order, trying to lay the groundwork

VarenyaJ committed:
Writing things a little out of order, trying to lay the groundwork. Still not ready

VarenyaJ committed:
Forgot to import Report class

VarenyaJ committed:
Setting up Steps 0 and 1 again

VarenyaJ committed:
Rename some variables for clarity. Continue with steps 2-4

VarenyaJ committed:
Edited 2 debug print statements

VarenyaJ committed:
Edited 2 debug print statements

VarenyaJ committed:
Rewrite step 5

VarenyaJ committed:
Write up a draft of Steps 6 and 7

VarenyaJ committed:
Write up a rough draft of Steps 6 and 7

VarenyaJ committed:
Pass only three directories to 'create_phenopacket_dataset'

VarenyaJ committed:
Pass only three paths to 'create_phenopacket_dataset'

VarenyaJ committed:
Fix typo

VarenyaJ committed:
Add comment about potentially over-aggressive deduplication

VarenyaJ committed:
Still adjusting so this writes one phenopacket per patient described in each PMID PDF. Adjust some variable names and comments

VarenyaJ committed:
Add HPO-specific prompt and Phenopacket-specific prompt

VarenyaJ committed:
Add missing quotation marks

VarenyaJ committed:
Fix typo

VarenyaJ committed:
Adjust variable name to reflect input type

VarenyaJ committed:
Use itertuples

VarenyaJ committed:
Fix critical error in scripts readme.md

VarenyaJ committed:
Rewrite command in README to work from project root

VarenyaJ committed:
Make adjustments to the intake CSV now that Step 5 of scripts/readme.md` is fixed

VarenyaJ committed:
Quick push

VarenyaJ committed:
Update Markdown cell

VarenyaJ committed:
Try to improve caching of clinical texts

VarenyaJ committed:
Cell 4 failed :(

VarenyaJ committed:
Try modifying to build the text cache for each UNIQUE PDF

VarenyaJ committed:
Suppress all those PyTorch/DataLoader warnings with a little `warnings.catch_warnings()` and then skip any PDFs that fail to convert, logging them and carrying on.

VarenyaJ committed:
Work on saving cache to disk

VarenyaJ committed:
Save the cache only once all the PDFs are parsed

VarenyaJ committed:
Added a hashtag so I could comment this error:
------------------------------------------------------------------------
ParseError                                Traceback (most recent call last)
File ~/miniconda3/envs/p5/lib/python3.13/site-packages/google/protobuf/json_format.py:544, in _Parser._ConvertFieldValuePair(self, js, message, path)
    543     continue
--> 544   raise ParseError(
    545       ('Message type "{0}" has no field named "{1}" at "{2}".\n'
    546        ' Available Fields(except extensions): "{3}"').format(
    547            message_descriptor.full_name, name, path,
    548            [f.json_name for f in message_descriptor.fields]))
    549 if name in names:

ParseError: Message type "org.phenopackets.schema.v2.Phenopacket" has no field named "cells" at "Phenopacket".
 Available Fields(except extensions): "['id', 'subject', 'phenotypicFeatures', 'measurements', 'biosamples', 'interpretations', 'diseases', 'medicalActions', 'files', 'metaData']"

During handling of the above exception, another exception occurred:

ParseError                                Traceback (most recent call last)
File ~/Desktop/Illini+Uni/Personalized_Genomic_Medicine-Precision_Genomics_Laboratory/PreGen/P5/notebooks/utils/phenopacket.py:44, in Phenopacket.__init__(self, phenopacket_json)
     43 try:
---> 44     _ = ParseDict(phenopacket_json, ProtoPhenopacket())
     45     feats = phenopacket_json["phenotypicFeatures"]

File ~/miniconda3/envs/p5/lib/python3.13/site-packages/google/protobuf/json_format.py:461, in ParseDict(js_dict, message, ignore_unknown_fields, descriptor_pool, max_recursion_depth)
    460 parser = _Parser(ignore_unknown_fields, descriptor_pool, max_recursion_depth)
--> 461 parser.ConvertMessage(js_dict, message, '')
    462 return message

File ~/miniconda3/envs/p5/lib/python3.13/site-packages/google/protobuf/json_format.py:502, in _Parser.ConvertMessage(self, value, message, path)
    501 else:
--> 502   self._ConvertFieldValuePair(value, message, path)
    503 self.recursion_depth -= 1

File ~/miniconda3/envs/p5/lib/python3.13/site-packages/google/protobuf/json_format.py:629, in _Parser._ConvertFieldValuePair(self, js, message, path)
    628   else:
--> 629     raise ParseError(str(e))
    630 except ValueError as e:

ParseError: Message type "org.phenopackets.schema.v2.Phenopacket" has no field named "cells" at "Phenopacket".
 Available Fields(except extensions): "['id', 'subject', 'phenotypicFeatures', 'measurements', 'biosamples', 'interpretations', 'diseases', 'medicalActions', 'files', 'metaData']"

During handling of the above exception, another exception occurred:

InvalidPhenopacketError                   Traceback (most recent call last)
Cell In[39], line 35
     32 list_input_texts.append(clinical_text)
     34 # Load the ground-truth Phenopacket
---> 35 truth_packet = Phenopacket.load_from_file(truth_path)
     36 list_truth_packets.append(truth_packet)
     38 # Extract per-patient IDs from raw JSON

File ~/Desktop/Illini+Uni/Personalized_Genomic_Medicine-Precision_Genomics_Laboratory/PreGen/P5/notebooks/utils/phenopacket.py:90, in Phenopacket.load_from_file(cls, path)
     88     data = json.load(f)
     89 logger.info("Loaded JSON data from %s", path)
---> 90 return cls(data)

File ~/Desktop/Illini+Uni/Personalized_Genomic_Medicine-Precision_Genomics_Laboratory/PreGen/P5/notebooks/utils/phenopacket.py:48, in Phenopacket.__init__(self, phenopacket_json)
     46 except (TypeError, ParseError) as e:
     47     logger.error("Failed to validate phenopacket: %s", e, exc_info=True)
---> 48     raise InvalidPhenopacketError(f"Failed to validate phenopacket: {e}")
     50 # If we reach this step then phenopacket_json is guaranteed to be valid, so we can successfully cache the raw JSON and the features list
     51 self._json = phenopacket_json

InvalidPhenopacketError: Failed to validate phenopacket: Message type "org.phenopackets.schema.v2.Phenopacket" has no field named "cells" at "Phenopacket".
 Available Fields(except extensions): "['id', 'subject', 'phenotypicFeatures', 'measurements', 'biosamples', 'interpretations', 'diseases', 'medicalActions', 'files', 'metaData']"

VarenyaJ committed:
Re-enable deduplication

VarenyaJ committed:
Small clarification

VarenyaJ committed:
Fix operation where we drop duplicates but we never printed the before count, so there was no way to verify it actually ran. Also the `print` call for “after deduplication” might be malformed to show a tuple rather than formatted text. Change Step 1 to track  both original and post-dedup counts to confirm dedup actually happened. Then use an f-string to try to avoid the tuple‐print bug.

VarenyaJ committed:
Try starting a mechanism to only intake 100 PDFs at a time

VarenyaJ committed:
use `os` for system paths

VarenyaJ committed:
Add discovery of `phenopacket-store`

VarenyaJ committed:
Adjust and simplify some code in the PDF conversion step

VarenyaJ committed:
Modify Step 4 with the new implementations

VarenyaJ committed:
Adjusted deduplication

VarenyaJ committed:
Adjusted discovery, expand `skipped_pdfs` to store reasons for skipping over errors

VarenyaJ committed:
Fixed parentheses typo

VarenyaJ committed:
Remove redundant cache and enforce the maximum limit of clinical inputs as well schema enforcement to prevent unknown field errors

VarenyaJ committed:
Fix stray errors

VarenyaJ committed:
Some quick quality changes

VarenyaJ committed:
attempt rolling back to 05d00c1ea8851311960a454eeaa1a94c3a2ed4cd on protected branch

VarenyaJ committed:
Pull changes from Main (git pull --no-rebase origin main): Update README

VarenyaJ committed:
Add ipykernel for python notebook

VarenyaJ committed:
Make the scripts use relative imports

VarenyaJ committed:
Revise `scripts` help commands

VarenyaJ committed:
Change help command to download 10 PDFs

VarenyaJ committed:
Try writing a main files to run all the modules together

VarenyaJ committed:
Update `scripts` readme.md

VarenyaJ committed:
Update `scripts` readme.md

VarenyaJ committed:
Update `scripts` readme.md and ``main.py

VarenyaJ committed:
Update `scripts` readme.md

VarenyaJ committed:
Update `scripts` readme.md

VarenyaJ committed:
Work on making this project structured like a package

VarenyaJ committed:
Change “python -m” invocations and the help-strings

VarenyaJ committed:
Refactor tests

VarenyaJ committed:
Changes to the package

VarenyaJ committed:
Indentation fix

VarenyaJ committed:
Small README header because I am getting confused switching back-and-forth

VarenyaJ committed:
Fix typo

VarenyaJ committed:
Update README

VarenyaJ committed:
Refactoring the notebook

VarenyaJ committed:
Actually use the `scripts` modules

VarenyaJ committed:
Step 0 no longer breaks

VarenyaJ committed:
Add TODO

VarenyaJ committed:
Update TODO

VarenyaJ committed:
Refactor Phenopacket parsing and caching, replace non-existent `from_dict` calls, and add input validation.

- Stop using `Phenopacket.from_dict(...)`, instead pass the parsed JSON directly into `Phenopacket(...)`, joining LLM output lines to ensure valid JSON.
- Factor out MD5 hashing into a new `compute_pdf_hash` helper and introduce `save_caches()` to persist both the phenopacket JSON cache and the PDF text pickle cache after each case.
- Ensure the ground-truth notebooks directory is created in setup, and have the dataset loader print and verify every PDF input path.
- Add a runtime guard to abort with a clear error if no clinical texts are loaded.
- Improve error handling and logging throughout the PDF loading loop for better clarity and resilience.

VarenyaJ committed:
Attempt fix at LLM inferencing

VarenyaJ committed:
Add `Any` from `typing` and catch `PdfiumError`

VarenyaJ committed:
Small markdown change

VarenyaJ committed:
Checking directories

VarenyaJ committed:
Clarify subsetting

VarenyaJ committed:
No need to prevent importing the text_cache

VarenyaJ committed:
Remove excess caching logic

VarenyaJ committed:
Cache Integrity & Versioning

VarenyaJ committed:
Move new caching implementation

VarenyaJ committed:
Fix cache creation

VarenyaJ committed:
Forgot colon

VarenyaJ committed:
Error with limiting sample selection: fix from type annotation to proper dict type annotation

VarenyaJ committed:
Error with limiting sample selection: fix from type annotation to proper dict type annotation. Fix some comments

VarenyaJ committed:
Build a strict system+user conversation

VarenyaJ committed:
Feed just the JSON‐array slice instead of feeding the whole raw output

VarenyaJ committed:
Adjust error raising

VarenyaJ committed:
Trying to still fix the prompting

VarenyaJ committed:
Re-arrange imports

VarenyaJ committed:
Mild change

VarenyaJ committed:
Add ruff format

VarenyaJ committed:
Mini-experiment: write an experimental cell to try to use the Ollama HTTP endpoint port while enforcing strict return via LLM output. Spoiler: does not work

VarenyaJ committed:
Moved non-functional experimental cell to end of notebook

VarenyaJ committed:
add note to the requirements folder in case someone encounters the common error of having Anaconda's repositories blocked

VarenyaJ committed:
Remove expired hashing from Step 2 and rework Step 4.5 (still borked)

VarenyaJ committed:
Remove large jumbled notebook

VarenyaJ committed:
Update gitignore

VarenyaJ committed:
Start Notebook from beginning; Refactor Project to become installable via 'pip install -e .'

VarenyaJ committed:
Update CI

VarenyaJ committed:
Update CI

VarenyaJ commented:
Hey @SmartMonkey-git could you advise on what the problem with the CI is?

VarenyaJ committed:
Just use Black for formatting, keep Ruff for Linting

VarenyaJ committed:
Formatting adjustments

VarenyaJ committed:
Fix CI

VarenyaJ committed:
Fix CI

SmartMonkey-git commented on pyproject.toml:
```diff
@@ -0,0 +1,49 @@
+where = ["src"]
+
+
+# Omit the [project.dependencies] section entirely to leave all dependency management to Conda.
```
Are you sure this works? How does setuptools know what packages to install?

SmartMonkey-git commented on requirements/environment.yml:
```diff
@@ -1,12 +1,15 @@
+      - -r requirements.txt         # core LLM & converter libraries (ollama, llama-index, docling)
+      - -r requirements_scripts.txt # CLI helpers like click
+      - -r requirements_test.txt    # testing & formatting (pytest, black)
+      - -e ..                        # install the package in editable mode
```
Why is that needed?

VarenyaJ replied:
Is it not? I thought it should be added as a pip command during setup. If not I can remove it

SmartMonkey-git replied:
You added it, so you should know if its needed or not. :D If its needed let me know why ^^

SmartMonkey-git commented on requirements/instructions_for_charite_conda_block.md:
```diff
@@ -0,0 +1,59 @@
+### Check for Blockages
```
Didnt you want to share this in teams?

VarenyaJ replied:
Yes but I had just moved it here while I watch if there are more Conda errors

SmartMonkey-git replied:
Did you find any other errors?

VarenyaJ replied:
Not yet. Encountered an issue where a test failed if Ollama was not running but that was 1) the first time and 2) a separate issue

VarenyaJ replied:
Shared it on the `Coding` channel in Teams

SmartMonkey-git commented on requirements/requirements.txt:
```diff
@@ -3,3 +3,5 @@ docling==2.36.1
 phenopackets==2.0.2.post4
 protobuf==3.20.3
 scikit-learn==1.7.0
+jupyterlab==4.4.4
+ipykernel==6.29.5
```
Dont think thats needed for the package.

VarenyaJ replied:
For the notebook

SmartMonkey-git replied:
Ok, but if you distribute this later on pypi and i do `pip install p5` how do I run the notebooks?

SmartMonkey-git replied:
the requirement.txt should only have requirements for whats in src.

VarenyaJ replied:
Where would you suggest I move them to?

SmartMonkey-git replied:
Scripts or tests. Although, they do not really fit in any.

VarenyaJ replied:
Moved it to `requirements_scripts.txt`

SmartMonkey-git commented on src/P5/__init__.py:
```diff
@@ -0,0 +1 @@
+__version__ = "0.1.0"
```
There is something called `setuptools_scm` that you can use to automat versioning

VarenyaJ replied:
I'll look into that! Thanks for the recommendation. As of now this isn't a big priority so I'm not too focused on that fix

SmartMonkey-git commented on src/P5/scripts/__main__.py:
```diff
@@ -0,0 +1,97 @@
+from P5.scripts.create_phenopacket_dataset import create_phenopacket_dataset
+
+
+@click.group(invoke_without_command=True)
```
I dont see this script or any of the others mentioned in the pyproject tools. For promotizer i had to do something like:

```
[project.scripts]
promotizer = "promotizer.cli:cli"
```
Where the cli was in src/promotizer/cli.py and the name of the cli is cli.

VarenyaJ replied:
Is this not correct?:
`pyproject.toml`:
```
[project.scripts]
pull-git-files              = "P5.scripts.pull_git_files:pull_git_files"
create-pmid-pkl             = "P5.scripts.create_pmid_pkl:create_pmid_pkl"
pmid-downloader             = "P5.scripts.pmid_downloader:pmid_downloader"
create-phenopacket-dataset  = "P5.scripts.create_phenopacket_dataset:create_phenopacket_dataset"
file-to-phenopacket         = "P5.scripts.file_to_phenopacket:file_to_phenopacket"
```


Would this be the correct revision?:
```
[project.scripts]
p5 = "P5.scripts.__main__:cli"
```

SmartMonkey-git commented on src/P5/scripts/readme.md:
```diff
@@ -0,0 +1,118 @@
+# P5/scripts - README
```
I think having a sphinx documentation makes more sense than writing all these readmes. Make sure you can port them later on

VarenyaJ replied:
Yes I agree. I think I'll eventually move to having a site like `https://p5.github.io/P5/`, but for now it's not the focus

VarenyaJ committed:
Attempt on making this package pip-installable - still unresolved

VarenyaJ committed:
Add IFGC keywords

VarenyaJ committed:
Add Miniconda and Mamba Notes relevant for setup

VarenyaJ committed:
skip data folder in tests directory

VarenyaJ committed:
black formatting

VarenyaJ committed:
Add to README

VarenyaJ committed:
Try black formatting again

VarenyaJ committed:
Continue to rework this into a package

VarenyaJ committed:
Continue to rework this into a package

VarenyaJ committed:
Black and Ruff formatting

VarenyaJ committed:
P5: add `pdf-parse` subcommand + robust help; keep P6 untouched

- New CLI: `p5 pdf-parse` (stub)
  - Enumerates PDFs/office/txt under an input path
  - Writes manifest.jsonl; optional text dump for .txt
  - Output organized under out/run_YYYY-MM-DD_HH-MM-SS/
- Register subcommand on the main Click group
- Harden help UX: `p5`, `p5 --help`, and `python -m P5 --help` all print help
- No changes in P6

VarenyaJ committed:
P5: wire `pdf-parse` to docling with robust fallbacks

- Use docling to extract text from PDF/DOCX/PPTX/HTML (when installed)
- Keep TXT passthrough; record manifest.jsonl with engine and output paths
- Save extracts under out/run_YYYY-MM-DD_HH-MM-SS/text/
- Graceful error handling: write .err files and continue
- New option: --engine {auto,docling} (default: auto)

VarenyaJ committed:
Delete top-level main file

VarenyaJ committed:
P5: fix ruff issues, simplify __main__, and refactor pdf-parse

- Remove stray top-level __main__.py (use src/P5/__main__.py only)
- __main__.py: clean help UX, register pdf-parse, no unused imports
- pdf-parse: narrow exception handling (no blind except), split into helper to reduce complexity, keep docling auto-detect
- scripts/__main__.py: remove undefined `p5` arg
- Formatting via black / ruff

VarenyaJ committed:
P5: pdf-parse — add .txt passthrough + docling integration + CLI polish

- Always passthrough .txt (don’t send to docling)
- Clear status line: “docling (txt passthrough)”
- Write manifest.jsonl and extracted text under out/run_YYYY-MM-DD_HH-MM-SS/
- Code formatted (black) and lint clean (ruff)

VarenyaJ committed:
Add exclusions

VarenyaJ committed:
black --skip-magic-trailing-comma .

SmartMonkey-git commented on pyproject.toml:
```diff
@@ -0,0 +1,82 @@
+# Extras mirror our requirements/*.txt so `pip install .[llm,scripts,test]` works too.
+dependencies = []
+
+[project.optional-dependencies]
+llm = ["ollama==0.5.1", "docling==2.36.1", "phenopackets==2.0.2.post4", "protobuf==3.20.3", "scikit-learn==1.7.0", "jupyterlab==4.4.4", "ipykernel==6.29.5"]
+scripts = ["click==8.2.1", "GitPython==3.1.44", "requests==2.32.4", "biopython==1.85", "selenium==4.33.0", "pandas==2.2.3"]
+test = ["black==25.1.0", "pytest==8.4.0", "ruff==0.12.1"]
```
You should not have two sources of truth for packages.
Either pyproject.toml or requirements.txt

VarenyaJ replied:
Still working on trying to streamline this

VarenyaJ replied:
I am using the `pyproject.toml ` for pip an keeping the `require*.txt` for conda installs at the moment

SmartMonkey-git replied:
You can look at the template project, to see how you can use the requirements as single source of truth.

VarenyaJ committed:
remove excess setup files

VarenyaJ committed:
Revise package: switch to Hatch + scripts CLI; remove setuptools files

VarenyaJ committed:
package cli: unify CLI around `p5`, deprecate `python -m P5`; simplify __main__ and expose pdf-parse as its own entry point

- pyproject.toml:
  - set `p5` entry point to `P5.scripts.__main__:cli` (the full scripts CLI)
  - added `p5-pdf-parse` console script alias entry for users to run, giving direct access to the pdf-parse helper
  - kept optional sub-tool entry points (pull-git-files, create-pmid-pkl, etc.)
  - adjusted keywords and comments for clarity

- src/P5/__main__.py:
  - replaced the old standalone click group with a deprecation shim
  - running `python -m P5` now prints a friendly message directing users to `p5`

VarenyaJ committed:
Add a `Quickstart` pip section to the README.md; revise other sections

VarenyaJ committed:
update gitignore

VarenyaJ committed:
Fix capitalization

VarenyaJ committed:
Add TOOO: reference template repo for how to use the requirements as single source of truth

VarenyaJ committed:
feat(step1): load dataset CSV, validate columns, deduplicate PMIDs, and preview

VarenyaJ committed:
chore(step2): enumerate phenopacket-store JSONs for visibility and sanity

VarenyaJ committed:
feat(step3): add docling PDF→text conversion with persistent cache

VarenyaJ committed:
feat(step4): load clinical texts and validate ground-truth phenopackets via project utils

VarenyaJ committed:
feat(step5): add strict HPO extraction prompt, schema, robust parsing, and wrapping util

VarenyaJ committed:
test(step6): run single-case HPO extraction and build minimal predicted phenopacket

VarenyaJ committed:
feat(step7): batch-extract HPOs, save raw LLM output and minimal predicted phenopackets

VarenyaJ committed:
feat(step8): evaluate predictions via PhenotypeEvaluator and generate Report

VarenyaJ committed:
feat(step9): persist final evaluation Report to JSON

VarenyaJ committed:
docs(optional): add commented semantic similarity skeleton for future ontology-based metrics

VarenyaJ committed:
Finish adding skeleton code for revising the demonstration notebook

VarenyaJ committed:
Finish adding skeleton code for revising the demonstration notebook

VarenyaJ committed:
fix(cli): make file_to_phenopacket and pmid_downloader more robust in the event of an HTPP 500 error:

- file_to_phenopacket:
  • Always request JSON from Ollama (`format="json"`)
  • Fallback: if model output is not valid JSON, write a minimal valid
    Phenopacket scaffold so one JSON file is produced per input
  • Ensure output directory exists before writing
  • Added RFC3339 timestamp helper for `metaData.created`

- pmid_downloader:
  • Hardened _get_pmcid() with retry/backoff for transient NCBI 5xx errors
  • Return None (skip PMID) instead of crashing on Entrez/HTTP errors
  • Added Entrez.email env override + graceful handling of failures

These changes ensure:
- `test_file_to_phenopacket[.pdf]` produces expected JSON outputs
- `test_pmid_downloader` no longer fails on NCBI server errors
- CLI commands always exit 0 with sensible output, even in edge cases

VarenyaJ committed:
Add comments and docstrings while dealing with the HTTP 500 error

VarenyaJ committed:
Add comments and docstrings while dealing with the HTTP 500 error

VarenyaJ committed:
[WIP] notebooks/demo.ipynb: clean up metadata, improve robustness, and upgrade model while still revising

- Added stable `id` fields for all notebook cells to improve reproducibility
- Step 4: added `list_pmids_loaded` to track actually loaded PMIDs, updated patient ID fallback
- Step 5: improved minimal Phenopacket builder
  - skip invalid HPOs
  - ensure non-null labels
  - use UTC ISO8601 timestamps in `metaData.created`
- Temporarily upgraded model from `llama3.2:latest` → `gpt-oss:latest`
- Removed deprecated `--hidethinking` option; added `num_ctx` for larger context
- Step 6 & 7: use `list_pmids_loaded` instead of indexing into dataframe
- Step 8: updated evaluator `model` field to `ollama:gpt-oss:latest`

VarenyaJ committed:
Downgrade from gpt-oss:latest

VarenyaJ committed:
First partial success of Demo: rerun notebook with new PMIDs + refined HPO extraction

- Added debug print of input char counts per PMID.
- Relaxed HPO JSON schema (require only `hpo_id`).
- Rewrote HPO prompt instructions for onset/severity/frequency + excerpts.
- Added fallback when schema-based JSON parse fails.
- Updated execution counts & timestamps.

VarenyaJ committed:
(configuration) CI: Use Ruff instead of Black

VarenyaJ committed:
Ruff check and formatting

VarenyaJ committed:
Update `.gitignore` to skip the generated files

VarenyaJ committed:
Move `jupyterlab==4.4.4` and `ipykernel==6.29.5` from `requirements.txt` to `requirements_scripts.txt` as per @SmartMonkey-git's request

VarenyaJ committed:
Rename and edit notes on Charité blocking conda

VarenyaJ committed:
Edit and rename the notes on conda blockages again
